datumaro/cli/commands/diff.py

```diff
@@ -2,6 +2,206 @@
+
+    equality_parser = parser.add_argument_group("Equality comparison options")
+    equality_parser.add_argument('-iia', '--ignore-item-attr', action='append',
+        help="Ignore item attribute (repeatable)")
+    equality_parser.add_argument('-ia', '--ignore-attr', action='append',
+        help="Ignore annotation attribute (repeatable)")
+    equality_parser.add_argument('-if', '--ignore-field', action='append',
+        help="Ignore annotation field (repeatable, default: %s)" % \
+            eq_default_if)
+    equality_parser.add_argument('--match-images', action='store_true',
+        help='Match dataset items by image pixels instead of ids')
+    equality_parser.add_argument('--all', action='store_true',
+        help="Include matches in the output")
+
+    return parser
+
+@error_rollback('on_error', implicit=True)
+def diff_command(args):
+    dst_dir = args.dst_dir
+    if dst_dir:
+        if not args.overwrite and osp.isdir(dst_dir) and os.listdir(dst_dir):
+            raise CliException("Directory '%s' already exists "
+                "(pass --overwrite to overwrite)" % dst_dir)
+    else:
+        dst_dir = generate_next_file_name('diff')
+    dst_dir = osp.abspath(dst_dir)
+
+    if not osp.exists(dst_dir):
+        on_error.do(rmtree, dst_dir, ignore_errors=True)
+        os.makedirs(dst_dir)
+
+    try:
+        project = load_project(args.project_dir)
+    except ProjectNotFoundError as e:
+        if args.project_dir:
+            raise
+        else:
+            project = None
+
+    try:
+        if not args.second_target:
+            first_dataset = project.working_tree.make_dataset()
+            second_dataset = parse_full_revpath(args.first_target, project)
+        else:
+            first_dataset = parse_full_revpath(args.first_target, project)
+            second_dataset = parse_full_revpath(args.second_target, project)
+    except Exception as e:
+        raise CliException(str(e))
+
+
+    if args.method is ComparisonMethod.equality:
+        if args.ignore_field:
+            args.ignore_field = eq_default_if
+        comparator = ExactComparator(
+            match_images=args.match_images,
+            ignored_fields=args.ignore_field,
+            ignored_attrs=args.ignore_attr,
+            ignored_item_attrs=args.ignore_item_attr)
+        matches, mismatches, a_extra, b_extra, errors = \
+            comparator.compare_datasets(first_dataset, second_dataset)
+
+        output = {
+            "mismatches": mismatches,
+            "a_extra_items": sorted(a_extra),
+            "b_extra_items": sorted(b_extra),
+            "errors": errors,
+        }
+        if args.all:
+            output["matches"] = matches
+
+        output_file = osp.join(dst_dir,
+            generate_next_file_name('diff', ext='.json', basedir=dst_dir))
+        log.info("Saving diff to '%s'" % output_file)
+        with open(output_file, 'w') as f:
+            json.dump(output, f, indent=4, sort_keys=True)
```

Comment:
The description for the `--output-dir` option says that when it's not specified, it means "don't save"; so shouldn't this code be conditional?

Reply:
Fixed arg description